File: tusow/src/main/java/it/unibo/coordination/tusow/api/UsersApi.java
package it.unibo.coordination.tusow.api;

import io.vertx.core.Promise;
import io.vertx.ext.web.RoutingContext;
import it.unibo.coordination.tusow.presentation.Link;
import it.unibo.coordination.tusow.presentation.User;

import java.util.Collection;

public interface UsersApi extends Api {

    void createUser(User userData, Promise<Link> promise);

    void readAllUsers(Integer skip, Integer limit, String filter, Promise<Collection<? extends User>> promise);

    void readUser(String identifier, Promise<User> promise);

    void updateUser(String identifier, User newUserData, Promise<User> promise);

    static UsersApi get(RoutingContext context) {
        throw new UnsupportedOperationException("not implemented");
    }
}

File: tusow-service/src/main/kotlin/it/unibo/coordination/tusow/api/UsersApi.kt
package it.unibo.coordination.tusow.api

import io.vertx.core.Promise
import io.vertx.ext.web.RoutingContext
import it.unibo.coordination.tusow.presentation.Link
import it.unibo.coordination.tusow.presentation.User

interface UsersApi : Api {
    fun createUser(userData: User, promise: Promise<Link>)
    fun readAllUsers(skip: Int, limit: Int, filter: String, promise: Promise<Collection<User>>)
    fun readUser(identifier: String, promise: Promise<User>)
    fun updateUser(identifier: String, newUserData: User, promise: Promise<User>)

    companion object {
        @Suppress("UNUSED_PARAMETER")
        @JvmStatic
        operator fun get(context: RoutingContext): UsersApi {
            throw UnsupportedOperationException("not implemented")
        }
    }
}